File: my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode : public rclcpp::Node 
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client")
    {   // 1 & 4 are the input for the service. The two ints we want to sum.
        thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4));
        // creating new threads inside the thread pool to call service multiple times
        // threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 61, 4)));
        // threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 31, 4)));
        // threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 21, 4)));

    }

    void callAddTwoIntsService(int a, int b)
    {
        
        // creating client node
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        // wait for the server to be ready
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(),"Waiting for the server to be up...");
        }
        // creating the request
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;
        // send the request asynchronously from the client to the server
        auto future_obj = client->async_send_request(request);

        try
        {   // this will pause the current thread at this point. spin(node) would not get processed 
            // The idea is to create another thread and keep it running
            auto response = future_obj.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %d", (int)a, (int)b, (int)response->sum);
        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
        
    }


private:
    std::thread thread1_;
    // to create a thread pool, uncomment the following line and delete the previous one:
    // std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(1, 5)
        # calling the server multiple times
        self.call_add_two_ints_server(1, 25)
        self.call_add_two_ints_server(1, 45)
        
    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server...")
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future_obj = client.call_async(request)
        future_obj.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))
    
    def callback_call_add_two_ints(self, future_obj, a, b):
        try:
            response = future_obj.result()
            self.get_logger().info(str(a) + " + " + str(b) + " = " + str(response.sum)) 

        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))

def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode() 
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()